contracts/xastro_token/src/contract.rs

```diff
@@ -0,0 +1,1308 @@
+use cw20_base::contract::{
+    execute_update_marketing, execute_upload_logo, query_download_logo, query_marketing_info,
+    query_minter, query_token_info,
+};
+use cw20_base::enumerable::query_all_allowances;
+use cw20_base::msg::ExecuteMsg;
+use cw20_base::state::{MinterData, TokenInfo, TOKEN_INFO};
+use cw20_base::ContractError;
+use cw_storage_plus::Bound;
+
+use astroport::xastro_token::{InstantiateMsg, QueryMsg};
+
+/// Contract name that is used for migration.
+const CONTRACT_NAME: &str = "astroport-xastro-token";
+/// Contract version that is used for migration.
+const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+// settings for pagination
+const MAX_LIMIT: u32 = 30;
+const DEFAULT_LIMIT: u32 = 10;
+
+/// ## Description
+/// Creates a new contract with the specified parameters in the [`InstantiateMsg`].
+/// Returns the default object of type [`Response`] if the operation was successful,
+/// or a [`ContractError`] if the contract was not created.
+/// ## Params
+/// * **deps** is the object of type [`DepsMut`].
+///
+/// * **_env** is the object of type [`Env`].
+///
+/// * **_info** is the object of type [`MessageInfo`].
+/// * **msg** is a message of type [`InstantiateMsg`] which contains the basic settings for creating a contract.
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+    mut deps: DepsMut,
+    env: Env,
+    _info: MessageInfo,
+    msg: InstantiateMsg,
+) -> StdResult<Response> {
+    set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+
+    // Check valid token info
+    msg.validate()?;
+
+    // Create initial accounts
+    let total_supply = create_accounts(&mut deps, &env, msg.initial_balances.as_slice())?;
+
+    capture_total_supply_history(deps.storage, &env, total_supply)?;
+
+    // Check supply cap
+    if let Some(limit) = msg.get_cap() {
+        if total_supply > limit {
+            return Err(StdError::generic_err("Initial supply greater than cap"));
+        }
+    }
+
+    let mint = match msg.mint {
+        Some(m) => Some(MinterData {
+            minter: deps.api.addr_validate(&m.minter)?,
+            cap: m.cap,
+        }),
+        None => None,
+    };
+
+    // Store token info
+    let data = TokenInfo {
+        name: msg.name,
+        symbol: msg.symbol,
+        decimals: msg.decimals,
+        total_supply,
+        mint,
+    };
+
+    TOKEN_INFO.save(deps.storage, &data)?;
+
+    Ok(Response::default())
+}
+
+pub fn create_accounts(deps: &mut DepsMut, env: &Env, accounts: &[Cw20Coin]) -> StdResult<Uint128> {
+    let mut total_supply = Uint128::zero();
+
+    for row in accounts {
+        let address = deps.api.addr_validate(&row.address)?;
+        BALANCES.save(deps.storage, &address, &row.amount, env.block.height)?;
+        total_supply += row.amount;
+    }
+
+    Ok(total_supply)
+}
+
+/// ## Description
+/// Default CW20 execute method implementation.
+#[cfg_attr(not(feature = "library"), entry_point)]
```

Comment:
I think we should add a complete description for each function. To be able to generate documentation for Cargo in the future.

Reply:
Ok, done